src/common/buffer.cc

```diff
@@ -800,53 +836,175 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
+                *((uint64_t*)(c+sizeof(uint64_t))) = *((uint64_t*)(p+sizeof(uint64_t)));
+                return _len + _off;
+            case 8:
+                *((uint64_t*)(c)) = *((uint64_t*)(p));
+                return _len + _off;
+            case 4:
+                *((uint32_t*)(c)) = *((uint32_t*)(p));
+                return _len + _off;
+            case 2:
+                *((uint16_t*)(c)) = *((uint16_t*)(p));
+                return _len + _off;
+            case 1:
+                *((uint8_t*)(c)) = *((uint8_t*)(p));
+                return _len + _off;
+        }
+        int cursor = 0;
+        while (l >= sizeof(uint64_t)) {
+            *((uint64_t*)(c + cursor)) = *((uint64_t*)(p + cursor));
+            cursor += sizeof(uint64_t);
+            l -= sizeof(uint64_t);
+        }
+        while (l >= sizeof(uint32_t)) {
+            *((uint32_t*)(c + cursor)) = *((uint32_t*)(p + cursor));
+            cursor += sizeof(uint32_t);
+            l -= sizeof(uint32_t);
+        }
+        while (l > 0) {
+            *(c+cursor) = *(p+cursor);
+            cursor++;
+            l--;
+        }
+    }
+    else {
+        memcpy(c, p, l);
+        _len += l;
+    }
+    return _len + _off;
   }
-    
-  void buffer::ptr::copy_in(unsigned o, unsigned l, const char *src)
+  
+  void buffer::ptr::copy_in(unsigned o, unsigned l, const char *src, bool crc_reset)
   {
     assert(_raw);
     assert(o <= _len);
     assert(o+l <= _len);
-    _raw->invalidate_crc();
-    memcpy(c_str()+o, src, l);
+    char* dest = _raw->data + _off + o;
+    if (crc_reset)
+        _raw->invalidate_crc();
+    if (l < 64) {
+        switch (l) {
+            case 1:
+                *((uint8_t*)(dest)) = *((uint8_t*)(src));
+                return;
+            case 2:
+                *((uint16_t*)(dest)) = *((uint16_t*)(src));
+                return;
+            case 3:
+                *((uint16_t*)(dest)) = *((uint16_t*)(src));
+                *((uint8_t*)(dest+2)) = *((uint8_t*)(src+2));
+                return;
+            case 4:
+                *((uint32_t*)(dest)) = *((uint32_t*)(src));
+                return;
+            case 8:
+                *((uint64_t*)(dest)) = *((uint64_t*)(src));
+                return;
+            default:
+                int cursor = 0;
+                while (l >= sizeof(uint64_t)) {
+                    *((uint64_t*)(dest + cursor)) = *((uint64_t*)(src + cursor));
+                    cursor += sizeof(uint64_t);
+                    l -= sizeof(uint64_t);
+                }
+                while (l >= sizeof(uint32_t)) {
+                    *((uint32_t*)(dest + cursor)) = *((uint32_t*)(src + cursor));
+                    cursor += sizeof(uint32_t);
+                    l -= sizeof(uint32_t);
+                }
+                while (l > 0) {
+                    *(dest + cursor) = *(src + cursor);
+                    cursor++;
+                    l--;
+                }
+                return;
```

Comment:
I guess we could abstract the same logic with buffer::ptr::append and buffer::ptr::copy_out